app_flutter/lib/status_grid.dart

```diff
@@ -143,24 +146,94 @@ class StatusGridHelper {
   }
 
   /// Create [List<Task>] for [List<TaskIcon>] at the top of [StatusGrid].
-  List<Task> _createTaskIconRow(
-      List<CommitStatus> statuses, Map<String, int> taskColumnKeyIndex) {
-    final List<Task> taskIconRow = List<Task>(taskColumnKeyIndex.keys.length);
-
-    for (int statusIndex = 0; statusIndex < statuses.length; statusIndex++) {
-      final CommitStatus status = statuses[statusIndex];
-      for (Stage stage in status.stages) {
-        for (Task task in stage.tasks) {
-          final int taskIndex = taskColumnKeyIndex[_taskColumnKey(task)];
-          if (taskIconRow[taskIndex] == null) {
-            taskIconRow[taskIndex] = task;
-          }
+  ///
+  /// Tasks are organized by column in [matrix].
+  List<Task> _createTaskIconRow(List<List<Task>> matrix) {
+    final List<Task> taskIconRow = List<Task>(matrix[0].length);
+
+    // In the worst case, this has to scan the entire matrix to build the first row. However,
+    // the common in production is that the first row has the task.
+    for (int column = 0; column < matrix[0].length; column++) {
+      for (int row = 0; row < matrix.length; row++) {
+        if (matrix[row][column] != null) {
+          taskIconRow[column] = matrix[row][column];
+          break;
         }
       }
     }
 
     return taskIconRow;
   }
+
+  /// Sort [columnKeyIndex] based on a list of weights.
+  /// 
+  /// For duplicate weights, order is assumed to not matter.
+  @visibleForTesting
+  static Map<String, int> sortColumnKeyIndex(
+      Map<String, int> columnKeyIndex, List<int> weights) {
+    final Map<String, int> sortedColumnKeyIndex = <String, int>{};
+
+    // Map the current index to its given weight
+    final Map<int, int> weightIndex = <int, int>{};
+    for (int i = 0; i < weights.length; i++) {
+      weightIndex[i] = weights[i];
+    }
+
+    // Sort by weight in [weightIndex]
+    final List<int> sortedWeightKeys = weightIndex.keys.toList()
+      ..sort((int k1, int k2) => weightIndex[k1].compareTo(weightIndex[k2]));
+    final LinkedHashMap<int, int> sortedWeights =
+        LinkedHashMap<int, int>.fromIterable(sortedWeightKeys,
+            key: (dynamic k) => k, value: (dynamic k) => weightIndex[k]);
+
+    // Reassign indices based on the order
+    int newIndex = 0;
+    sortedWeights.forEach((int key, int value) {
```

Comment:
I'm a bit confused by this part. You have the sorted map of indexes/weights, you should be able to use that to look things instead of traversing the entire columnKeyIndex

Reply:
I realized creating a "reverse" map allowed for the quick lookup. I was able to simplify this further, thank you for pointing it out!